File: transcript_formatter/services/tie/correction_object.py
"""CorrectionObject contract + validator (ATIA §4.8, D8.1 & D8.4).

The CorrectionObject is the atomic unit of transcript intelligence: one proposed
change against the immutable canonical Deepgram baseline, produced by a
deterministic rule or an AI specialty prompt, and reviewed by a human reporter.

The JSON Schema at ``schema/correction_object.schema.json`` is the single source
of truth. This module loads that schema to derive its enums / required fields, so
the backend can never silently drift from the contract the frontend also consumes.
The conditional per-``change.type`` constraints (the schema's ``allOf``/``if``/
``then`` blocks) are enforced explicitly here — a text correction must carry
before/after; a structural correction must carry ``structural_change`` and must
NOT carry before/after.

This module has zero third-party dependencies (no pydantic / jsonschema), so it
runs anywhere the pipeline runs.
"""

from __future__ import annotations

import json
import re
import secrets
import time
from pathlib import Path
from typing import Any, Iterable

# ── schema as single source of truth ─────────────────────────────────────────

_SCHEMA_PATH = Path(__file__).resolve().parents[2] / "schema" / "correction_object.schema.json"


def _load_schema() -> dict:
    return json.loads(_SCHEMA_PATH.read_text(encoding="utf-8"))


_SCHEMA = _load_schema()
_PROPS = _SCHEMA["properties"]

REQUIRED_FIELDS: tuple[str, ...] = tuple(_SCHEMA["required"])
SPECIALTIES: frozenset[str] = frozenset(_PROPS["specialty"]["enum"])
CHANGE_TYPES: frozenset[str] = frozenset(_PROPS["change"]["properties"]["type"]["enum"])
REASON_KINDS: frozenset[str] = frozenset(_PROPS["reason_kind"]["enum"])
PROVENANCE_SOURCES: frozenset[str] = frozenset(
    _PROPS["provenance"]["properties"]["source"]["enum"]
)
REVIEW_STATES: frozenset[str] = frozenset(_PROPS["review"]["properties"]["state"]["enum"])

_ID_PATTERN = re.compile(_PROPS["id"]["pattern"])
_REASON_MIN = _PROPS["reason"]["minLength"]
_REASON_MAX = _PROPS["reason"]["maxLength"]
_CONF_MIN = _PROPS["confidence"]["minimum"]
_CONF_MAX = _PROPS["confidence"]["maximum"]

# Derived from the schema's change.allOf branches (kept in sync with §4.8).
TEXT_CHANGE_TYPES: frozenset[str] = frozenset(
    {"proper_name_correction", "medical_term_correction", "contextual_number_flag"}
)
STRUCTURAL_CHANGE_TYPES: frozenset[str] = frozenset(
    {
        "speaker_reassignment",
        "qa_split",
        "objection_attribution",
        "examination_section_change",
        "off_record_boundary_mark",
    }
)

_LOCATION_REQUIRED = ("paragraph_id", "start_word_id", "end_word_id")

# Generic reasons the validator rejects (ATIA §2.4: reasons must be specific).
_GENERIC_REASONS = {"improved clarity", "looks better", "better", "clarity", "fix", "correction"}


class CorrectionValidationError(ValueError):
    """Raised when a correction payload violates the CorrectionObject contract.

    Carries every problem found (not just the first) so callers can surface a
    complete diagnosis."""

    def __init__(self, errors: Iterable[str]) -> None:
        self.errors = list(errors)
        super().__init__("; ".join(self.errors) or "invalid correction")


def new_correction_id() -> str:
    """Return a ``corr_``-prefixed ULID matching the schema's id pattern.

    Crockford base32, time-ordered (48-bit ms timestamp + 80 bits randomness).
    Uses stdlib ``time``/``secrets`` — deterministic ordering is not required for
    correctness, only monotonic-ish sortability."""
    alphabet = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"  # Crockford; all within [0-9A-Z]
    timestamp = int(time.time() * 1000) & ((1 << 48) - 1)
    value = (timestamp << 80) | secrets.randbits(80)
    chars = []
    for _ in range(26):
        chars.append(alphabet[value & 0x1F])
        value >>= 5
    return "corr_" + "".join(reversed(chars))


def _is_number(value: Any) -> bool:
    return isinstance(value, (int, float)) and not isinstance(value, bool)


def validate_correction(data: Any) -> None:
    """Validate a correction payload against the CorrectionObject contract.

    Raises ``CorrectionValidationError`` (with every problem found) on any
    violation; returns ``None`` when valid."""
    errors: list[str] = []

    if not isinstance(data, dict):
        raise CorrectionValidationError(["correction must be an object"])

    for field in REQUIRED_FIELDS:
        if field not in data:
            errors.append(f"missing required field '{field}'")

    _validate_id(data.get("id"), errors)
    _validate_enum(data.get("specialty"), SPECIALTIES, "specialty", errors)
    _validate_reason(data.get("reason"), errors)
    _validate_enum(data.get("reason_kind"), REASON_KINDS, "reason_kind", errors)
    _validate_confidence(data.get("confidence"), errors)
    _validate_location(data.get("location"), errors)
    _validate_change(data.get("change"), errors)
    _validate_provenance(data.get("provenance"), errors)
    _validate_review(data.get("review"), errors)

    if errors:
        raise CorrectionValidationError(errors)


# ── field validators ──────────────────────────────────────────────────────

def _validate_id(value: Any, errors: list[str]) -> None:
    if value is None:
        return  # missing-required already reported
    if not isinstance(value, str) or not _ID_PATTERN.match(value):
        errors.append(f"id must match {_ID_PATTERN.pattern} (got {value!r})")


def _validate_enum(value: Any, allowed: frozenset[str], field: str, errors: list[str]) -> None:
    if value is None:
        return
    if value not in allowed:
        errors.append(f"{field} must be one of {sorted(allowed)} (got {value!r})")


def _validate_reason(value: Any, errors: list[str]) -> None:
    if value is None:
        return
    if not isinstance(value, str):
        errors.append("reason must be a string")
        return
    if not (_REASON_MIN <= len(value) <= _REASON_MAX):
        errors.append(f"reason length must be {_REASON_MIN}-{_REASON_MAX} chars (got {len(value)})")
    if value.strip().lower() in _GENERIC_REASONS:
        errors.append(f"reason is too generic: {value!r}")


def _validate_confidence(value: Any, errors: list[str]) -> None:
    if value is None:
        return
    if not _is_number(value) or not (_CONF_MIN <= value <= _CONF_MAX):
        errors.append(f"confidence must be a number in [{_CONF_MIN}, {_CONF_MAX}] (got {value!r})")


def _validate_location(value: Any, errors: list[str]) -> None:
    if value is None:
        return
    if not isinstance(value, dict):
        errors.append("location must be an object")
        return
    for key in _LOCATION_REQUIRED:
        if not value.get(key):
            errors.append(f"location.{key} is required")
    for offset_key in ("start_offset_ms", "end_offset_ms"):
        if offset_key in value and not (isinstance(value[offset_key], int) and value[offset_key] >= 0):
            errors.append(f"location.{offset_key} must be a non-negative integer")


def _validate_change(value: Any, errors: list[str]) -> None:
    if value is None:
        return
    if not isinstance(value, dict):
        errors.append("change must be an object")
        return

    change_type = value.get("type")
    if change_type is None:
        errors.append("change.type is required")
        return
    if change_type not in CHANGE_TYPES:
        errors.append(f"change.type must be one of {sorted(CHANGE_TYPES)} (got {change_type!r})")
        return

    has_before = "before" in value and value["before"] is not None
    has_after = "after" in value and value["after"] is not None
    has_structural = isinstance(value.get("structural_change"), dict)

    if change_type in TEXT_CHANGE_TYPES:
        if not has_before or not has_after:
            errors.append(f"change.type '{change_type}' requires both 'before' and 'after'")
    elif change_type in STRUCTURAL_CHANGE_TYPES:
        if not has_structural:
            errors.append(f"change.type '{change_type}' requires 'structural_change'")
        if has_before or has_after:
            errors.append(f"change.type '{change_type}' must not carry 'before'/'after' text")
    # inconsistency_flag: marks a location; no before/after/structural required.


def _validate_provenance(value: Any, errors: list[str]) -> None:
    if value is None:
        return
    if not isinstance(value, dict):
        errors.append("provenance must be an object")
        return
    if not value.get("generated_at"):
        errors.append("provenance.generated_at is required")
    source = value.get("source")
    if source is None:
        errors.append("provenance.source is required")
    else:
        _validate_enum(source, PROVENANCE_SOURCES, "provenance.source", errors)
        if source == "ai" and not value.get("provider"):
            errors.append("provenance.provider is required when source == 'ai'")


def _validate_review(value: Any, errors: list[str]) -> None:
    if value is None:
        return
    if not isinstance(value, dict):
        errors.append("review must be an object")
        return
    state = value.get("state")
    if state is None:
        errors.append("review.state is required")
    else:
        _validate_enum(state, REVIEW_STATES, "review.state", errors)

File: src/lib/transcript/correctionObject.ts
// TypeScript side of the CorrectionObject contract (ATIA §4.8, D8.1).
//
// Contract of record: the co-located JSON Schema
// src/lib/transcript/correction_object.schema.json (surviving governed home,
// relocated from the retiring Python transcript_formatter/schema/ per DOC-0327).
// This module is the SURVIVING PRODUCTION AUTHORITY — the only CorrectionObject
// definition on a live path (frontend + ai-review Edge Function). The schema is
// kept in lockstep by a drift test (correctionObjectSchema.test.ts) that fails if
// the enums/required diverge, so the two are mechanically synchronized rather than
// independently authored. The Python validator (services/tie/correction_object.py)
// derives from the same schema and retires with transcript_formatter/.
// The per-type rules below are the schema's change.allOf branches.
//
// A CorrectionObject is one proposed change against the immutable Deepgram
// baseline: produced by an AI (bridge/specialty) prompt or a deterministic rule,
// reviewed by a human court reporter.

export type CorrectionSpecialty =
  | "proper_name_novel"
  | "medical_context"
  | "qa_split"
  | "speaker_reassignment"
  | "objection_attribution"
  | "examination_section"
  | "off_record_boundary"
  | "inconsistency_flag"
  | "contextual_number"
  | "phonetic_disambiguation"
  | "deterministic_rule";

export type CorrectionChangeType =
  | "proper_name_correction"
  | "medical_term_correction"
  | "speaker_reassignment"
  | "qa_split"
  // Structural EXTRACTION of an embedded objection into its own unit (DOC-0325
  // Wave 4 / Decision B). Deliberately SEPARATE from objection_attribution: this
  // type owns only the split boundary (the objection word span); attribution
  // (which attorney objected) is a distinct concern that may be resolved or left
  // unidentified. structural_change: {objection_start_word_id, objection_end_word_id,
  // objector_speaker_id?}. Never carries a fabricated speaker.
  | "objection_split"
  | "objection_attribution"
  | "examination_section_change"
  | "off_record_boundary_mark"
  | "inconsistency_flag"
  | "contextual_number_flag";

export type ReasonKind =
  | "registry_match"
  | "phonetic_similarity"
  | "context_pattern"
  | "prior_correction"
  | "inconsistency_detected"
  | "structural_boundary"
  | "confidence_threshold"
  | "reporter_preference"
  | "rule_pattern";

export type ReviewState = "pending" | "accepted" | "rejected" | "edited" | "superseded";
export type ProvenanceSource = "ai" | "deterministic" | "reporter";

// change.type sets that gate before/after vs structural_change (schema allOf).
export const TEXT_CHANGE_TYPES: ReadonlySet<CorrectionChangeType> = new Set([
  "proper_name_correction",
  "medical_term_correction",
  "contextual_number_flag",
]);
export const STRUCTURAL_CHANGE_TYPES: ReadonlySet<CorrectionChangeType> = new Set([
  "speaker_reassignment",
  "qa_split",
  "objection_split",
  "objection_attribution",
  "examination_section_change",
  "off_record_boundary_mark",
]);

export interface CorrectionLocation {
  paragraph_id: string;
  start_word_id: string;
  end_word_id: string;
  start_offset_ms?: number;
  end_offset_ms?: number;
}

export interface CorrectionChange {
  type: CorrectionChangeType;
  before?: string;
  after?: string;
  structural_change?: Record<string, unknown>;
}

export interface CorrectionProvenance {
  source: ProvenanceSource;
  provider?: string;
  model?: string;
  prompt_versions?: Record<string, string>;
  context_hash?: string;
  generated_at: string;
}

export interface CorrectionReview {
  state: ReviewState;
  decided_by?: string | null;
  decided_at?: string | null;
  decision_note?: string | null;
  final_value?: Record<string, unknown> | null;
}

export type SupportingEvidenceKind =
  | "registry_entry"
  | "deepgram_confidence"
  | "prior_transcript_correction"
  | "glossary_entry"
  | "speaker_pattern_match"
  | "phonetic_similarity_score";

export interface CorrectionSupportingEvidence {
  kind: SupportingEvidenceKind;
  id?: string;
  value?: string | number;
  case_id?: string;
}

export interface CorrectionDownstream {
  applied_to_working_transcript?: boolean;
  applied_at?: string | null;
  reverted_at?: string | null;
  // Set when review.state is accepted/edited but applied_to_working_transcript is
  // false: accepted, apply deferred (e.g. qa_split awaiting the structural apply
  // engine). Workspace shows "accepted — apply pending" rather than a silent no-op.
  pending_reason?: string | null;
}

export interface CorrectionObject {
  id: string;
  transcript_id: string;
  case_id: string;
  specialty: CorrectionSpecialty;
  prompt_version: string;
  location: CorrectionLocation;
  change: CorrectionChange;
  reason: string;
  reason_kind: ReasonKind;
  confidence: number;
  confidence_source?: string;
  provenance: CorrectionProvenance;
  supporting_evidence?: CorrectionSupportingEvidence[];
  review: CorrectionReview;
  downstream: CorrectionDownstream;
}

const SPECIALTIES = new Set<string>([
  "proper_name_novel", "medical_context", "qa_split", "speaker_reassignment",
  "objection_attribution", "examination_section", "off_record_boundary",
  "inconsistency_flag", "contextual_number", "phonetic_disambiguation", "deterministic_rule",
]);
const CHANGE_TYPES = new Set<string>([...TEXT_CHANGE_TYPES, ...STRUCTURAL_CHANGE_TYPES, "inconsistency_flag"]);
const REASON_KINDS = new Set<string>([
  "registry_match", "phonetic_similarity", "context_pattern", "prior_correction",
  "inconsistency_detected", "structural_boundary", "confidence_threshold",
  "reporter_preference", "rule_pattern",
]);
const PROVENANCE_SOURCES = new Set<string>(["ai", "deterministic", "reporter"]);
const REVIEW_STATES = new Set<string>(["pending", "accepted", "rejected", "edited", "superseded"]);

const ID_PATTERN = /^corr_[0-9A-Z]{26}$/;
const REASON_MIN = 10;
const REASON_MAX = 500;
const GENERIC_REASONS = new Set(["improved clarity", "looks better", "better", "clarity", "fix", "correction"]);

export class CorrectionValidationError extends Error {
  errors: string[];
  constructor(errors: string[]) {
    super(errors.join("; ") || "invalid correction");
    this.name = "CorrectionValidationError";
    this.errors = errors;
  }
}

// Returns the list of contract violations ([] when valid). Mirrors the Python
// validator in services/tie/correction_object.py.
export function collectCorrectionErrors(data: unknown): string[] {
  const errors: string[] = [];
  if (!data || typeof data !== "object") {
    return ["correction must be an object"];
  }
  const c = data as Record<string, unknown>;

  if (typeof c.id !== "string" || !ID_PATTERN.test(c.id)) {
    errors.push("id must match ^corr_[0-9A-Z]{26}$");
  }
  for (const field of ["transcript_id", "case_id", "prompt_version"]) {
    if (typeof c[field] !== "string" || !(c[field] as string)) errors.push(`${field} is required`);
  }
  if (!SPECIALTIES.has(c.specialty as string)) errors.push(`specialty invalid: ${String(c.specialty)}`);
  if (!REASON_KINDS.has(c.reason_kind as string)) errors.push(`reason_kind invalid: ${String(c.reason_kind)}`);

  const reason = c.reason;
  if (typeof reason !== "string") {
    errors.push("reason must be a string");
  } else {
    if (reason.length < REASON_MIN || reason.length > REASON_MAX) {
      errors.push(`reason length must be ${REASON_MIN}-${REASON_MAX} (got ${reason.length})`);
    }
    if (GENERIC_REASONS.has(reason.trim().toLowerCase())) errors.push(`reason is too generic: ${reason}`);
  }

  const conf = c.confidence;
  if (typeof conf !== "number" || Number.isNaN(conf) || conf < 0 || conf > 1) {
    errors.push(`confidence must be a number in [0,1] (got ${String(conf)})`);
  }

  errors.push(...locationErrors(c.location));
  errors.push(...changeErrors(c.change));
  errors.push(...provenanceErrors(c.provenance));
  errors.push(...reviewErrors(c.review));
  errors.push(...downstreamErrors(c.downstream));
  return errors;
}

export function validateCorrection(data: unknown): asserts data is CorrectionObject {
  const errors = collectCorrectionErrors(data);
  if (errors.length > 0) throw new CorrectionValidationError(errors);
}

function locationErrors(loc: unknown): string[] {
  if (!loc || typeof loc !== "object") return ["location must be an object"];
  const l = loc as Record<string, unknown>;
  const errs: string[] = [];
  for (const key of ["paragraph_id", "start_word_id", "end_word_id"]) {
    if (!l[key]) errs.push(`location.${key} is required`);
  }
  return errs;
}

function changeErrors(change: unknown): string[] {
  if (!change || typeof change !== "object") return ["change must be an object"];
  const ch = change as Record<string, unknown>;
  const type = ch.type as string;
  if (!CHANGE_TYPES.has(type)) return [`change.type invalid: ${String(type)}`];

  const errs: string[] = [];
  const hasBefore = ch.before != null;
  const hasAfter = ch.after != null;
  const hasStructural = ch.structural_change != null && typeof ch.structural_change === "object";

  if (TEXT_CHANGE_TYPES.has(type as CorrectionChangeType)) {
    if (!hasBefore || !hasAfter) errs.push(`change.type '${type}' requires both 'before' and 'after'`);
  } else if (STRUCTURAL_CHANGE_TYPES.has(type as CorrectionChangeType)) {
    if (!hasStructural) errs.push(`change.type '${type}' requires 'structural_change'`);
    if (hasBefore || hasAfter) errs.push(`change.type '${type}' must not carry 'before'/'after' text`);
  }
  return errs;
}

function provenanceErrors(prov: unknown): string[] {
  if (!prov || typeof prov !== "object") return ["provenance must be an object"];
  const p = prov as Record<string, unknown>;
  const errs: string[] = [];
  if (!p.generated_at) errs.push("provenance.generated_at is required");
  if (!PROVENANCE_SOURCES.has(p.source as string)) {
    errs.push(`provenance.source invalid: ${String(p.source)}`);
  } else if (p.source === "ai" && !p.provider) {
    errs.push("provenance.provider is required when source == 'ai'");
  }
  return errs;
}

function reviewErrors(review: unknown): string[] {
  if (!review || typeof review !== "object") return ["review must be an object"];
  const r = review as Record<string, unknown>;
  if (!REVIEW_STATES.has(r.state as string)) return [`review.state invalid: ${String(r.state)}`];
  return [];
}

// downstream is a schema-required object (the Python validator enforces it via the
// schema's `required` list). TS previously skipped it, so a payload missing
// downstream passed TS but failed Python/schema — this closes that divergence.
function downstreamErrors(downstream: unknown): string[] {
  if (!downstream || typeof downstream !== "object") return ["downstream must be an object"];
  return [];
}

// corr_ + 26-char Crockford base32 ULID (48-bit ms timestamp + 80 bits random).
// Uses Web Crypto — available in both the browser and the Deno Edge runtime.
const _CROCKFORD = "0123456789ABCDEFGHJKMNPQRSTVWXYZ";

export function newCorrectionId(now: number = Date.now()): string {
  const rand = new Uint8Array(10);
  crypto.getRandomValues(rand);
  let value = (BigInt(now) & ((1n << 48n) - 1n)) << 80n;
  for (const byte of rand) value = (value << 8n) | BigInt(byte);
  let out = "";
  for (let i = 0; i < 26; i += 1) {
    out = _CROCKFORD[Number(value & 31n)] + out;
    value >>= 5n;
  }
  return "corr_" + out;
}
